README.md
# Habitica Automation

[![](https://github.com/believer/habitica/workflows/Release/badge.svg)](https://github.com/believer/habitica/actions?workflow=Release)

I got inspired by @alexanderczigler's [automation scripts](https://github.com/alexanderczigler/habitica) for our [Habitica](http://habitica.com/) party and wanted to try out Google Apps Scripts for myself.

## Get started

This project is written in TypeScript and uses `clasp` to integrate and build to Google Apps Scripts. Start by installing `clasp` and logging in.

```bash
$ npm i -g @google/clasp
$ clasp login
```

Get your API values from the [Habitica settings](https://habitica.com/user/settings/api) and create a `config.ts` in the root of the project with the following content:

```typescript
// config.ts
const config = {
  goldThresholds: {
    // If the user has more than 500 gold, buy an armoire
    armoire: 500,
  },
  manaThresholds: {
    // If the user has more that 150 mana, cast earthquake
    earthquake: 150,
  },
  headers: {
    'x-api-user': '<user-id>',
    'x-api-key': '<api-key>',
  },
}

export default config
```

You should the be ready to push your project to Google Apps Scripts using.
`clasp` will take care of compiling the TypeScript code.

```bash
$ clasp push
```

## Functions

### `buyArmoire`

Buy an "Enchanted Armoire"

### `earthquake`

Cast "Earthquake" that buffs the party's intelligence

### `hatchPets`

Hatch any eggs you have laying around

### `feedPets`

Feed those hungry pets

### `joinQuest`

Join a pending quest


CHANGELOG.md
## [1.4.5](https://github.com/believer/habitica/compare/v1.4.4...v1.4.5) (2021-03-25)


### Bug Fixes

* **feedPets:** only feed to the maximum amount allowed ([0a0c007](https://github.com/believer/habitica/commit/0a0c00722c5e6544080c1da181d901a8a5d19256))

## [1.4.4](https://github.com/believer/habitica/compare/v1.4.3...v1.4.4) (2021-03-20)


### Bug Fixes

* **hatchPets:** handle no combinations found ([f9981ee](https://github.com/believer/habitica/commit/f9981eefffb536daa87f1d3a9466d250b4144359))

## [1.4.3](https://github.com/believer/habitica/compare/v1.4.2...v1.4.3) (2021-03-15)


### Bug Fixes

* **feedPets:** actually feed the specific food ([bf35b7a](https://github.com/believer/habitica/commit/bf35b7a1a36523dad3e60488a8ff95d4bed6655c))

## [1.4.2](https://github.com/believer/habitica/compare/v1.4.1...v1.4.2) (2021-03-15)


### Bug Fixes

* **feedPets:** handle pets which have mounts ([87011a6](https://github.com/believer/habitica/commit/87011a62799438291418b925ef1e0319f93bc910))

## [1.4.1](https://github.com/believer/habitica/compare/v1.4.0...v1.4.1) (2021-03-15)


### Bug Fixes

* handle special foods ([99645ff](https://github.com/believer/habitica/commit/99645ffcfeffaa97384ab0d39db8cc3cbef81d81))

# [1.4.0](https://github.com/believer/habitica/compare/v1.3.0...v1.4.0) (2021-03-11)


### Bug Fixes

* add tools of trade runner ([de212a9](https://github.com/believer/habitica/commit/de212a9daade9874fbbfd453b449c55b46eb2cd3))


### Features

* **spell:** add cast of tools of trade ([88f898e](https://github.com/believer/habitica/commit/88f898e27768e8fe22c505f8dd79b1d2e8427244))

# [1.3.0](https://github.com/believer/habitica/compare/v1.2.1...v1.3.0) (2021-03-10)


### Features

* **feedPets:** feed multiple of one item ([3285a4f](https://github.com/believer/habitica/commit/3285a4fbce66fd79b4b1b9c0ec73890eb261145f))

## [1.2.1](https://github.com/believer/habitica/compare/v1.2.0...v1.2.1) (2021-03-08)


### Bug Fixes

* **feedPets:** only feed valid pets ([07ecf33](https://github.com/believer/habitica/commit/07ecf33fba59c0e55a64ccb2deab17a960a798d1))

# [1.2.0](https://github.com/believer/habitica/compare/v1.1.0...v1.2.0) (2021-03-08)


### Features

* add healthPotion ([ac062e9](https://github.com/believer/habitica/commit/ac062e981154aa28bfe8105ca042fdd802e5f22a))

# [1.1.0](https://github.com/believer/habitica/compare/v1.0.1...v1.1.0) (2021-03-04)


### Bug Fixes

* add joinQuest runner ([7ced25d](https://github.com/believer/habitica/commit/7ced25d548b3a8145730af7d48d03d20a6672d26))


### Features

* add joinQuest ([207b4d7](https://github.com/believer/habitica/commit/207b4d70d7e05604392b2182690306244eb07264))

## [1.0.1](https://github.com/believer/habitica/compare/v1.0.0...v1.0.1) (2021-03-04)


### Bug Fixes

* **hatchPets:** handle missing potions and existing pet ([d8deb01](https://github.com/believer/habitica/commit/d8deb01fa6ad1783d35951903f086f7f892855e4))

# 1.0.0 (2021-03-04)


### Features

* add feedPets ([39c1bfd](https://github.com/believer/habitica/commit/39c1bfd2583a68e8d5e437491a362a7f756e53ab))
* add hatchPets ([3540b14](https://github.com/believer/habitica/commit/3540b145d04bdac8a486b0e167f8cb7ec4dffa5e))
* initial commit ([e148b1a](https://github.com/believer/habitica/commit/e148b1a8b6fdc0bb2298b0e7bf7618cd39e33a35))


__tests__/habitica.test.ts
import Habitica from '../habitica'

jest.mock('../config', () => ({
  manaThresholds: {
    earthquake: 150,
  },
  goldThresholds: {
    armoire: 500,
  },
  headers: 'headers',
}))

let UrlFetchApp: any
let Logger: any

beforeEach(() => {
  UrlFetchApp = {
    fetch: jest.fn(),
  }

  Logger = {
    log: jest.fn(),
  }

  global.UrlFetchApp = UrlFetchApp
  global.Logger = Logger
})

const setup = ({ user }) => {
  UrlFetchApp.fetch.mockImplementationOnce(() => ({
    getContentText: jest.fn(() =>
      JSON.stringify({
        data: user,
      })
    ),
  }))

  return new Habitica()
}

describe('#buyArmoire', () => {
  test('buy if gold is above threshold, get food', () => {
    const habitica = setup({
      user: {
        stats: {
          gp: 600,
        },
      },
    })

    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn(() =>
        JSON.stringify({
          data: {
            armoire: {
              type: 'food',
              dropText: 'fish',
            },
          },
        })
      ),
    }))

    habitica.buyArmoire()

    expect(Logger.log).toHaveBeenCalledWith('You gained fish')
  })

  test('buy if gold is above threshold, get item', () => {
    const habitica = setup({
      user: {
        stats: {
          gp: 600,
        },
      },
    })

    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn(() =>
        JSON.stringify({
          data: {
            armoire: {
              type: 'sword',
              dropText: 'fish',
              value: 1,
            },
          },
        })
      ),
    }))

    habitica.buyArmoire()

    expect(Logger.log).toHaveBeenCalledWith('You gained 1 sword')
  })

  test('buys nothing if gold is below threshold', () => {
    const habitica = setup({
      user: {
        stats: {
          gp: 400,
        },
      },
    })

    habitica.buyArmoire()

    expect(Logger.log).toHaveBeenCalledWith(
      'Wallet contains 400 coins. Threshold = 500'
    )
  })
})

describe('#earthquake', () => {
  test('cast if mana is above threshold', () => {
    const habitica = setup({
      user: {
        stats: {
          mp: 160,
        },
      },
    })

    // Cast quake
    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn().mockReturnValue('{ "success": true }'),
    }))

    habitica.earthquake()

    expect(UrlFetchApp.fetch).toHaveBeenCalledWith(
      'https://habitica.com/api/v3/user/class/cast/earth',
      {
        method: 'post',
        headers: 'headers',
      }
    )
    expect(Logger.log).toHaveBeenCalledWith('Casting *earthquake*')
  })

  test('does not cast if mana is below threshold', () => {
    const habitica = setup({
      user: {
        stats: {
          mp: 149,
        },
      },
    })

    // Cast quake
    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn().mockReturnValue('{ "success": true }'),
    }))

    habitica.earthquake()

    expect(Logger.log).toHaveBeenCalledWith('Mana is at 149. Threshold = 150')
  })
})

describe('#hatchPets', () => {
  test('hatches any available eggs', () => {
    const habitica = setup({
      user: {
        items: {
          eggs: {
            PandaCub: 1,
            BearCub: 0,
            Wolf: 0,
            Cactus: 0,
            Dragon: 0,
            TigerCub: 0,
            FlyingPig: 1,
            Fox: 0,
            LionCub: 0,
            SeaSerpent: 0,
          },
          hatchingPotions: {
            Skeleton: 1,
            Shade: 0,
            Red: 1,
            CottonCandyPink: 0,
            RoyalPurple: 0,
            CottonCandyBlue: 0,
            BlackPearl: 0,
            Desert: 1,
            Golden: 0,
            Base: 4,
            Zombie: 1,
            White: 1,
          },
        },
      },
    })

    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn().mockReturnValue('{ "success": true }'),
    }))
    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn().mockReturnValue('{ "success": true }'),
    }))

    habitica.hatchPets()

    expect(UrlFetchApp.fetch).toHaveBeenCalledWith(
      'https://habitica.com/api/v3/user/hatch/PandaCub/Skeleton',
      {
        method: 'post',
        headers: 'headers',
      }
    )
    expect(UrlFetchApp.fetch).toHaveBeenCalledWith(
      'https://habitica.com/api/v3/user/hatch/FlyingPig/Red',
      {
        method: 'post',
        headers: 'headers',
      }
    )

    expect(Logger.log).toHaveBeenCalledWith('Hatching Skeleton PandaCub')
    expect(Logger.log).toHaveBeenCalledWith('Hatching Red FlyingPig')
  })

  test('hatches nothing if no eggs', () => {
    const habitica = setup({
      user: {
        items: {
          eggs: {
            PandaCub: 0,
            BearCub: 0,
            Wolf: 0,
            Cactus: 0,
            Dragon: 0,
            TigerCub: 0,
            FlyingPig: 0,
            Fox: 0,
            LionCub: 0,
            SeaSerpent: 0,
          },
          hatchingPotions: {
            Skeleton: 1,
            Shade: 0,
            Red: 1,
            CottonCandyPink: 0,
            RoyalPurple: 0,
            CottonCandyBlue: 0,
            BlackPearl: 0,
            Desert: 1,
            Golden: 0,
            Base: 4,
            Zombie: 1,
            White: 1,
          },
        },
      },
    })

    habitica.hatchPets()

    expect(Logger.log).toHaveBeenCalledWith('All out of eggs')
  })
})

describe('#feedPets', () => {
  test('does not feed pets if no food is available', () => {
    const habitica = setup({
      user: {
        items: {
          pets: {
            'PandaCub-Skeleton': 27,
            'BearCub-Shade': 5,
            'Jackalope-RoyalPurple': 5,
            'Phoenix-Base': 5,
            'Wolf-Red': 30,
            'PandaCub-CottonCandyPink': 40,
            'BearCub-Skeleton': 5,
            'PandaCub-Shade': 30,
            'Cactus-CottonCandyBlue': 15,
            'Dragon-Skeleton': 5,
            'Dragon-Shade': 5,
            'Cactus-CottonCandyPink': 5,
            'Dragon-Red': 5,
            'PandaCub-CottonCandyBlue': 15,
            'TigerCub-Golden': 20,
            'FlyingPig-Base': 20,
            'Fox-Desert': 15,
            'TigerCub-Red': 5,
            'Cactus-Shade': 5,
            'LionCub-BlackPearl': 5,
            'FlyingPig-RoyalPurple': 5,
            'PandaCub-BlackPearl': 5,
            'LionCub-RoyalPurple': 5,
            'LionCub-Zombie': 25,
            'Fox-White': 15,
            'Dragon-BlackPearl': 5,
            'SeaSerpent-Golden': 5,
            'SeaSerpent-Red': 5,
            'SeaSerpent-Shade': 5,
          },
          food: {
            Chocolate: 0,
            CottonCandyPink: 0,
            Strawberry: 0,
            Fish: 0,
            Honey: 0,
            Meat: 0,
            CottonCandyBlue: 0,
            RottenMeat: 0,
            Potatoe: 0,
            Milk: 0,
          },
        },
      },
    })

    habitica.feedPets()

    expect(Logger.log).toHaveBeenCalledWith('All out of food')
  })

  test('feeds pets with available food', () => {
    const habitica = setup({
      user: {
        items: {
          pets: {
            'PandaCub-Skeleton': 27,
            'BearCub-Shade': 5,
            'Jackalope-RoyalPurple': 5,
            'Phoenix-Base': 5,
            'Wolf-Red': 30,
            'PandaCub-CottonCandyPink': 40,
            'BearCub-Skeleton': 5,
            'PandaCub-Shade': 30,
            'Cactus-CottonCandyBlue': 15,
            'Dragon-Skeleton': 5,
            'Dragon-Shade': 5,
            'Cactus-CottonCandyPink': 5,
            'Dragon-Red': 5,
            'PandaCub-CottonCandyBlue': 15,
            'TigerCub-Golden': 20,
            'FlyingPig-Base': 20,
            'Fox-Desert': 15,
            'TigerCub-Red': 5,
            'Cactus-Shade': 5,
            'LionCub-BlackPearl': 5,
            'FlyingPig-RoyalPurple': 5,
            'PandaCub-BlackPearl': 5,
            'LionCub-RoyalPurple': 5,
            'LionCub-Zombie': 25,
            'Fox-White': 15,
            'Dragon-BlackPearl': 5,
            'SeaSerpent-Golden': 5,
            'SeaSerpent-Red': 5,
            'SeaSerpent-Shade': 5,
          },
          food: {
            Chocolate: 0,
            CottonCandyPink: 0,
            Strawberry: 0,
            Fish: 1,
            Honey: 0,
            Meat: 2,
            CottonCandyBlue: 0,
            RottenMeat: 0,
            Potatoe: 0,
            Milk: 0,
          },
        },
      },
    })

    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn().mockReturnValue('{ "success": true }'),
    }))
    UrlFetchApp.fetch.mockImplementationOnce(() => ({
      getContentText: jest.fn().mockReturnValue('{ "success": true }'),
    }))

    habitica.feedPets()

    expect(UrlFetchApp.fetch).toHaveBeenCalledWith(
      'https://habitica.com/api/v3/user/feed/Phoenix-Base/Meat',
      {
        method: 'post',
        headers: 'headers',
      }
    )
    expect(UrlFetchApp.fetch).toHaveBeenCalledWith(
      'https://habitica.com/api/v3/user/feed/PandaCub-Skeleton/Fish',
      {
        method: 'post',
        headers: 'headers',
      }
    )

    expect(Logger.log).toHaveBeenCalledWith('Feeding Meat to Phoenix-Base')
    expect(Logger.log).toHaveBeenCalledWith('Feeding Fish to PandaCub-Skeleton')
  })
})


src/habitica.ts
import { HttpMethod, User, ArmoireResult, Party, Spell } from './types'
import config from './config'

const hasItems = (items: { [key: string]: number }) =>
  Object.entries(items)
    .filter(([_, value]) => value > 0)
    .flatMap(([item]) => item)

export default class Habitica {
  private baseUrl = 'https://habitica.com/api/v3'
  private config = config

  private fetch<T = any>(url: string, method: HttpMethod = HttpMethod.GET): T {
    const response = UrlFetchApp.fetch(`${this.baseUrl}${url}`, {
      method,
      headers: {
        'x-api-user': this.config.user.id,
        'x-api-key': this.config.user.key,
      },
    }).getContentText()
    const { data } = JSON.parse(response)

    return data
  }

  private getUser() {
    return this.fetch<User>('/user?userFields=stats,items')
  }

  private getParty() {
    return this.fetch<Party>('/groups/party')
  }

  private castSpell(spell: Spell, threshold: number) {
    const {
      stats: { mp: mana },
    } = this.getUser()

    if (mana > threshold) {
      this.fetch(`/user/class/cast/${spell}`, HttpMethod.POST)
      Logger.log(`Casting *${spell}*`)
    } else {
      Logger.log(`Mana is at ${mana}. Threshold = ${threshold}`)
    }
  }

  buyArmoire(): void {
    const {
      stats: { gp: gold },
    } = this.getUser()

    if (gold > this.config.goldThresholds.armoire) {
      const { armoire } = this.fetch<ArmoireResult>(
        '/user/buy-armoire',
        HttpMethod.POST
      )

      if (armoire.type === 'food') {
        Logger.log(`You gained ${armoire.dropText}`)
      } else {
        Logger.log(`You gained ${armoire.value} ${armoire.type}`)
      }
    } else {
      Logger.log(
        `Wallet contains ${gold} coins. Threshold = ${this.config.goldThresholds.armoire}`
      )
    }
  }

  earthquake(): void {
    this.castSpell(Spell.Earthquake, this.config.manaThresholds.earthquake)
  }

  toolsOfTrade(): void {
    this.castSpell(Spell.ToolsOfTrade, this.config.manaThresholds.toolsOfTrade)
  }

  hatchPets(): void {
    const { items } = this.getUser()

    const eggs = hasItems(items.eggs)

    if (eggs.length === 0) {
      Logger.log('All out of eggs')
      return
    }

    const hatchingPotions = hasItems(items.hatchingPotions)

    if (hatchingPotions.length === 0) {
      Logger.log('All out of potions')
      return
    }

    const pets = hasItems(items.pets)

    const combinations = eggs
      .map((egg) => hatchingPotions.map((potion) => `${egg}-${potion}`))
      .map(
        (combinedPets) =>
          combinedPets.filter((p) => !pets.some((pet) => pet === p))[0]
      )
      .filter(Boolean)

    combinations.forEach((combo) => {
      const [pet, egg] = combo.split('-')
      this.fetch(`/user/hatch/${pet}/${egg}`, HttpMethod.POST)
      Logger.log(`Hatching ${egg} ${pet}`)
    })
  }

  feedPets(): void {
    const { items } = this.getUser()

    const foods = Object.entries(items.food).filter(([_, value]) => value > 0)

    if (foods.length === 0) {
      Logger.log('All out of food')
    }

    const foodMap = {
      Base: 'Meat',
      CottonCandyBlue: 'CottonCandyBlue',
      CottonCandyPink: 'CottonCandyPink',
      Desert: 'Potatoe',
      Golden: 'Honey',
      Red: 'Strawberry',
      Shade: 'Chocolate',
      Skeleton: 'Fish',
      White: 'Milk',
      Zombie: 'RottenMeat',
    }

    const validPets = [
      'BearCub',
      'Cactus',
      'Dragon',
      'FlyingPig',
      'Fox',
      'LionCub',
      'PandaCub',
      'TigerCub',
      'Wolf',
    ]

    const pets = hasItems(items.pets)
      .map((pet) => pet.split('-'))
      .filter(([pet]) => validPets.includes(pet))
      .map(([pet, color]) => [`${pet}-${color}`, foodMap[color]])
      .filter(([pet]) => !items.mounts[pet])
      .filter(([_, food]) => food)

    const feed = foods
      .map(([food, value]) => {
        const matchFood = food.match(
          /(Base|CottonCandyBlue|CottonCandyPink|Desert|Golden|Red|Shade|Skeleton|White|Zombie)/gi
        )

        const petFood = matchFood?.length > 0 ? foodMap[matchFood[0]] : food
        const petThatLikesFood = pets.find(
          ([_, petLikes]) => petFood === petLikes
        )

        if (!petThatLikesFood) {
          return null
        }

        return [petThatLikesFood[0], food, value]
      })
      .filter(Boolean)

    feed.forEach(([pet, food, amount]) => {
      const maxFeed = Math.floor((50 - items.pets[pet]) / 5)
      const feedAmount = amount > 1 && amount > maxFeed ? maxFeed : amount

      this.fetch(
        `/user/feed/${pet}/${food}${
          feedAmount > 1 ? `?amount=${feedAmount}` : ''
        }`,
        HttpMethod.POST
      )
      Logger.log(`Feeding ${feedAmount} ${food} to ${pet}`)
    })
  }

  joinQuest(): void {
    const { quest } = this.getParty()

    if (!quest.key) {
      Logger.log('No current quest')
      return
    }

    if (quest.active) {
      Logger.log('Quest has already started')
      return
    }

    if (quest.members[this.config.user.id]) {
      Logger.log('You have already joined the quest')
      return
    }

    this.fetch('/groups/party/quests/accept', HttpMethod.POST)
    Logger.log('You have joined the quest! Happy hunting')
  }

  healthPotion(): void {
    const {
      stats: { hp, gp },
    } = this.getUser()

    if (hp > this.config.healthThresholds.health) {
      Logger.log('You are healthy enough')
      return
    }

    if (gp < 25) {
      Logger.log('Not enough gold to buy a health potion')
      return
    }

    this.fetch('/user/buy-health-potion', HttpMethod.POST)
    Logger.log(`You are healed up. Current hp = ${hp + 15}`)
  }
}
